### DISCUSSION TITLE
how to solve general nonlinear optimization probelms

### DISCUSSION THREAD

**Discussion by bc0428** [facebookresearch/theseus#651] [Q&A] [answered]
Hello, was wondering how to use the existing optimizers for inner optimization problems that are not in the form of least squares, such as nonlinear ones in general or perhaps convex ones.

Is it possible to integrate with optimizers in torch such as Adam, or is it necessary to implement a Theseus specific optimizer? 

I attempted to solve a convex problem with existing optimizers designed for least squares and the downstream performance is not ideal, which led me to suspect the optimizer has something to do with it

**Comment by luisenp** [comment, marked as answer]
Hi @bc0428. We have one optimizer that doesn't require the problem to be in the form of least squares, namely, [`th.DCEM`](https://github.com/facebookresearch/theseus/blob/main/theseus/optimizer/nonlinear/dcem.py#L38), which implements differentiable cross-entropy method ([paper](https://arxiv.org/abs/1909.12830)). To make the objective not codify a squared objective, you also need to change the default cost function aggregation function by passing a custom [error metric](https://github.com/facebookresearch/theseus/blob/main/theseus/core/objective.py#L32) to the [objective](https://github.com/facebookresearch/theseus/blob/main/theseus/core/objective.py#L69).  You can use the [implementation](https://github.com/facebookresearch/theseus/blob/main/theseus/core/objective.py#L37) of the default sum of squares  as a guide. 
 
Let me know if you have additional questions.

**Reply by bc0428** [reply]
thanks!!
